lib/grandpa/vote_message.go

```diff
@@ -236,22 +238,93 @@ func (s *Service) checkForEquivocation(voter *Voter, vote *SignedVote, stage Sub
 	if has {
 		// if the voter has already equivocated, every vote in that round is an equivocatory vote
 		eq[v] = append(eq[v], vote)
-		return true
+		return fmt.Errorf("%w: voter %s",
+			ErrEquivocation, v)
 	}
 
 	existingVote, has := s.loadVote(v, stage)
 	if !has {
-		return false
+		return nil
 	}
 
 	if has && existingVote.Vote.Hash != vote.Vote.Hash {
 		// the voter has already voted, all their votes are now equivocatory
 		eq[v] = []*SignedVote{existingVote, vote}
 		s.deleteVote(v, stage)
-		return true
+
+		err := s.reportEquivocation(stage, existingVote, vote)
+		if err != nil {
+			logger.Errorf("failed reporting equivocation: %s", err)
+		}
+		return fmt.Errorf("%w: voter %s has existing vote %s and new vote %s",
+			ErrEquivocation, v, existingVote.Vote.Hash, vote.Vote.Hash)
+	}
+
+	return nil
+}
+
+func (s *Service) reportEquivocation(stage Subround, existingVote *SignedVote, currentVote *SignedVote) error {
+	setID, err := s.grandpaState.GetCurrentSetID()
+	if err != nil {
+		return fmt.Errorf("getting authority set id: %w", err)
 	}
 
-	return false
+	round, err := s.grandpaState.GetLatestRound()
+	if err != nil {
+		return fmt.Errorf("getting latest round: %w", err)
+	}
+
+	pubKey := existingVote.AuthorityID
+
+	bestBlockHash := s.blockState.BestBlockHash()
+	runtime, err := s.blockState.GetRuntime(bestBlockHash)
+	if err != nil {
+		return fmt.Errorf("getting runtime: %w", err)
+	}
+
+	opaqueKeyOwnershipProof, err := runtime.GrandpaGenerateKeyOwnershipProof(setID, pubKey)
+	if err != nil {
+		return fmt.Errorf("getting key ownership proof: %w", err)
+	}
+
+	grandpaEquivocation := types.GrandpaEquivocation{
+		RoundNumber:     round,
+		ID:              pubKey,
+		FirstVote:       existingVote.Vote,
+		FirstSignature:  existingVote.Signature,
+		SecondVote:      currentVote.Vote,
+		SecondSignature: currentVote.Signature,
+	}
+
+	equivocationVote := types.NewGrandpaEquivocation()
+	switch stage {
+	case prevote:
+		err = equivocationVote.Set(types.PreVoteEquivocation(grandpaEquivocation))
+		if err != nil {
+			return fmt.Errorf("setting grandpa equivocation VDT as prevote equivocation: %w", err)
+		}
+	case precommit:
+		err = equivocationVote.Set(types.PreCommitEquivocation(grandpaEquivocation))
+		if err != nil {
+			return fmt.Errorf("setting grandpa equivocation VDT as precommit equivocation: %w", err)
+		}
+	case primaryProposal:
+		return fmt.Errorf("%w", errInvalidEquivocationStage)
+	default:
+		panic(fmt.Errorf("building equivocation proof: %s", errInvalidEquivocationStage))
```

Comment:
nit no need to error wrap + stack trace provide function name & line

```suggestion
		panic(fmt.Sprintf("equivocation stage not implemented: %s (%d)", stage, stage))
```